cryostatio/cryostat
# fix(rules): rule triggering on late-connecting targets

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #1494

## Description of the change:
This changeset:
1. handles `MODIFIED` JVM discovery events in a couple of places. Primarily to avoid an `UnsupportedOperationException` that was wrongly thrown and could cause rule activations to fail, and then secondarily to use these events to trigger a proactive re-check of non-connectable targets in case the discovery update made the target connectable.
2. Enhances the non-connectable target recheck logic. Previously these were always checked on a fixed 15 second timer. Now, the timer fires every 2 seconds, but each non-connectable target will only be rechecked according to some backoff timing logic, and after 1 minute of retry failures, that target will not be rechecked again.

## Motivation for the change:
This should fix the bug where rules may not activate against targets that appear but are not immediately connectable, and should also shorten the time it takes for Cryostat to detect that a slowly-connectable target becomes connectable.

## How to manually test:
Deploy test image in `crc` or other OpenShift/k8s. Deploy a sample application. Quite often, the k8s server notifies Cryostat of the new Endpoints object before the target JVM is actually ready to accept JMX connections, so the initial JVM ID retrieval attempt fails and so does the initial attempt to trigger automated rules. After this change, this workflow should again work as expected and Cryostat should notice that the target becomes connectable, and then triggers automated rules against it.
